File content is loaded into special variables

miqm opened the pull request:
# Contributing a Pull Request

If you haven't already, read the full [contribution guide](https://github.com/Azure/bicep/blob/main/CONTRIBUTING.md). The guide may have changed since the last time you read it, so please double-check. Once you are done and ready to submit your PR, run through the relevant checklist below.

## Contributing a feature

File loading functions are now loading files into special variables starting with `$` sign to avoid being overlapped by users. Then, in place where function is actually called, only variable access is emitted.
This will help to avoid duplicating file content when load functions are used in variables that need to be inlined.

If the loadTextContent or loadFileAsBase64 is used directly in a variable declaration (`var file = loadTextContent('file.txt')`) then no special symbol is emitted.


Fixes #5412

* [x] I have opened a new issue for the proposal, or commented on an existing one, and ensured that the Bicep maintainers are good with the design of the feature being implemented
* [x] I have included "Fixes #{issue_number}" in the PR description, so GitHub can link to the issue and close it when the PR is merged
* [x] I have appropriate test coverage of my new feature

majastrz commented on src/Bicep.Core/Visitors/FunctionVariableGeneratorVisitor.cs:
```diff
@@ -0,0 +1,51 @@
+                semanticModel.TypeManager.GetMatchedFunctionOverload(syntax) is { VariableGenerator: { } } functionOverload)
+            {
+                var variable = functionOverload.VariableGenerator(syntax, symbol, semanticModel.GetTypeInfo(syntax));
+                variables.Add(syntax, new InternalVariableSymbol($"$fxv#{variables.Count}", variable));
```
>fxv [](http://example.com/codeflow?start=68&length=3)

Why "fxv"? (No issue - just curious)

miqm replied:
Fx - function, v - variable

majastrz commented on src/Bicep.Core/Semantics/SemanticModel.cs:
```diff
@@ -154,6 +161,10 @@ public SemanticModel(Compilation compilation, BicepFile sourceFile, IFileResolve
+        /// <summary>
+        /// Gets variable symbols that are emitted by function calls. Used mostly to optimise loading content from files.
+        /// </summary>
+        public IImmutableDictionary<FunctionCallSyntaxBase, InternalVariableSymbol> InternalVariables => internalVariablesLazy.Value;
```
>public [](http://example.com/codeflow?start=8&length=6)

nit - needs new line after the property

majastrz commented on src/Bicep.Core/Emit/ExpressionConverter.cs:
```diff
@@ -842,7 +841,8 @@ private LanguageExpression ConvertVariableAccess(VariableAccessSyntax variableAc
                 default:
-                    throw new NotImplementedException($"Encountered an unexpected symbol kind '{symbol?.Kind}' when generating a variable access expression.");
+                    //just return a call to variables.
+                    return CreateFunction("variables", new JTokenExpression(name));                    
```
>CreateFunction [](http://example.com/codeflow?start=27&length=14)

default case should continue to throw, so it blows up when a new symbol is added.

miqm replied:
I created a special class called `ExplicitVariableAcccessSyntax` which is being detected in the ExpressionConverter and symbol check is being skipped for that C# type - we will always return `variable()` call for it.

majastrz commented on src/Bicep.Core/Semantics/InternalVariableSymbol.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace Bicep.Core.Semantics
+{
+    public class InternalVariableSymbol : Symbol
```
>Symbol [](http://example.com/codeflow?start=42&length=6)

This shouldn't be a symbol because it's not generated during binding and is not really part of the symbol hierarchy.

shenglol replied:
Agreed. I feel like the semantic model should not care about whether a special variable needs to be created or not, and this can be an optmization step during code generation (e.g., create a new visitor to visit all loadTextContent function calls and collect a list of variables we need to emit).

miqm replied:
changed to just `FunctionVariable`  POCO

miqm commented:
@majastrz @shenglol - I updated the PR, added also optimisation to not create special variables when function is used directly in variable declaration.

shenglol commented on src/Bicep.Core/Semantics/FunctionOverload.cs:
```diff
@@ -50,9 +52,14 @@ public FunctionOverload(string name, string genericDescription, string descripti
 
+        public VariableGeneratorDelegate? VariableGenerator { get; }
+
+
```
[nit]: Extra newline.

shenglol commented on src/Bicep.Core/Emit/ExpressionConverter.cs:
```diff
@@ -966,6 +971,7 @@ private LanguageExpression ConvertVariableAccess(VariableAccessSyntax variableAc
 
                 default:
                     throw new NotImplementedException($"Encountered an unexpected symbol kind '{symbol?.Kind}' when generating a variable access expression.");
+                  
```
[nit]: Extra newline.

shenglol commented on src/Bicep.Core/Semantics/SemanticModel.cs:
```diff
@@ -154,6 +161,11 @@ public SemanticModel(Compilation compilation, BicepFile sourceFile, IFileResolve
+        /// <summary>
+        /// Gets function variables that are emitted by function calls. Used mostly to optimise loading content from files.
+        /// </summary>
+        public IImmutableDictionary<FunctionCallSyntaxBase, FunctionVariable> FunctionVariables => functionVariablesLazy.Value;
```
Could we move this to the `EmitterContext` class, since this is an optmization for code gen?

miqm replied:
Good point, will move it.

miqm replied:
Moved.

shenglol commented on src/Bicep.Core/Visitors/FunctionVariableGeneratorVisitor.cs:
```diff
@@ -0,0 +1,66 @@
+            var variable = functionOverload.VariableGenerator(syntax, symbol, semanticModel.GetTypeInfo(syntax), directVariableAssignment);
+            if (variable is not null)
+            {
+                variables.Add(syntax, new($"$fxv#{variables.Count}", variable));
```
I wonder if it makes sense to change the variable name to some sort of combination of the function's fully qualified name and its parameter values (may need encoding). For instance, `"__sys.loadTextContent_abc.json__": "<content of abc.json>"`. The name can then be used to as a key to cache the variables to avoid generating multiple function variables with the same value. It also makes it possible to decompile a function variable into a function call in the future.

miqm replied:
I though about it but it would require to encode all function parameters in the variable name. Additionaly it's all about template size optimisation so creating long variable names is a bit against it. Also, I think it's going to be very rare to type load text content for same file multiple times in one template, usually developer will load file into single variable and ten reuse it. We could have a linter rule to suggest using single variable when this function is called multiple times with exact same arguments. But I'd like to stick with this auto-generated variable name.

shenglol replied:
Sounds good.

shenglol commented on src/Bicep.Core/Semantics/SemanticModel.cs:
```diff
@@ -315,5 +314,6 @@ private ImmutableArray<ResourceMetadata> GetAllResourceMetadata()
         public FileSymbol Root => this.Binder.FileSymbol;
 
         public ResourceScope TargetScope => this.Binder.TargetScope;
+
```
[nit]: Extra newline.

shenglol commented on src/Bicep.Core/Semantics/FunctionVariable.cs:
```diff
@@ -0,0 +1,19 @@
+
+namespace Bicep.Core.Semantics
+{
+    public class FunctionVariable
```
Mind moving this to Bicep.Core.Emit as well? Otherwise it looks good to me.

shenglol reviewed (approved):
:shipit:

majastrz commented on src/Bicep.Core/Emit/ExpressionConverter.cs:
```diff
@@ -926,8 +925,14 @@ private FunctionExpression GetLoopItemVariableExpression(ForSyntax @for, Languag
-            string name = variableAccessSyntax.Name.IdentifierName;
+            var name = variableAccessSyntax.Name.IdentifierName;
 
+            if (variableAccessSyntax is ExplicitVariableAccessSyntax)
```
>ExplicitVariableAccessSyntax [](http://example.com/codeflow?start=40&length=28)

I think we're forced into this approach because we don't have an intermediate representation between Bicep semantic model and the JSON.

I think we'll have to revisit this after we merge Anthony's IR PR.

majastrz commented on src/Bicep.Core/Syntax/ExplicitVariableAccessSyntax.cs:
```diff
@@ -0,0 +1,16 @@
+    /// <summary>
+    /// Represents an explicit reference to a variable
+    /// </summary>
+    public class ExplicitVariableAccessSyntax: VariableAccessSyntax
```
>ExplicitVariableAccessSyntax [](http://example.com/codeflow?start=17&length=28)

I do wonder if we should move this to an Emit-related namespace since this type of syntax node can't ever be produced by the parser.

majastrz reviewed (approved):
:shipit:
